cimr/processor/yamler.py

```diff
@@ -0,0 +1,370 @@
+        self.outdir = self.outdir_root + str(self.data_type) + '/'
+        pathlib.Path(self.outdir).mkdir(exist_ok=True)
+
+        if verify_weblink(self.file_link):
+            logging.info(f' starting download')
+            download_file(self.file_link, self.outdir)
+            self.hash = self.yaml_data['data_file']['location']['md5']
+            self.downloaded_file = self.outdir + self.downloaded_file
+        else:
+            logging.error(f' file unavailable')
+            sys.exit(1)
+
+
+    def make_archive_dir(self):
+        """Make a temp dir for downloaded files to be archived"""
+        self.archive_dir = 'submitted_data/downloaded_archive/'
+        pathlib.Path(self.archive_dir).mkdir(exist_ok=True)
+
+
+    def extract_bulk(self):
+        """Extract downloaded bulk file"""
+        import tarfile
+
+        if tarfile.is_tarfile(self.downloaded_file):
+            tarred_data = tarfile.open(
+                self.downloaded_file, 
+                mode='r:*'
+            )
+            # 'multiple' option is added for edge cases where multiple
+            # data type files in a tarfile share the same yaml
+            if self.data_type == 'multiple':
+                verify_dir(tarred_data)
+                tarred_data.extractall(path=self.outdir_root)
+            else:
+                for member in tarred_data.getmembers():
+                    if member.isreg():
+                        member.name = os.path.basename(member.name)
+                        tarred_data.extract(member, path=self.outdir)
+
+        else: # raise exception for invalid archive file
+            raise Exception(' invalid archive file for upload_bulk')
+
+        # moving the downloaded archive file to subdir
+        # 'downloaded_archive' to avoid being processed later
+        # ref: https://github.com/greenelab/cimr-d/pull/12  
+        self.make_archive_dir()
+        new_path = self.archive_dir + self.downloaded_file.split('/')[-1]
+        os.rename(
+            self.downloaded_file,
+            new_path
+        )
+        self.downloaded_file = new_path
+
+
+    def check_defined(self):
+        """Check whether the submitted data is a single file"""
+        if self.yaml_data['defined_as'] in ['upload', 'upload_bulk']:
+            self.download()
+        else:
+            logging.error(f' not an acceptible \'defined_as\' variable')
+            sys.exit(1)
+        
+        self.check_hash()
+
+        if self.yaml_data['defined_as'] == 'upload_bulk':
+            self.extract_bulk()
+    
+
+    def format_data(self):
+        """Initializing submitted column names to cimr variables"""
+        columnset = self.yaml_data['data_file']['columns']
+        reversekeys = {
+            v: k for k, v in columnset.items() if v != 'na'
+        }
+        # get a list of new submitted_data files to process
+        infile = self.file_link.split('/')[-1]
+        outfile = self.yaml_data['data_file']['output_name']
```

Comment:
It seems that `output_name` can easily mess up the whole setup. Do a user really need it?

Reply:
Is `output_name` a required field? If yes, you may need some error handling logic.

Reply:
Here is one test case that failed: 
Downloaded file is `foo.txt.gz`, output_name is `foo`, so `self.outfile_path` is same as its original location after it was downloaded. The downloaded file is silently moved to `submitted_data/downloaded_archive/`, and no file can be processed.